[BLOG] add post about multidimensional integration in scipy

ollybritton opened the pull request:
<!--Hi there! If you aren't submitting a blog post, you can delete this message.
Thank you for submitting a blog post! We want to make sure content produced for Quansight Labs is accessible, so please address the the following guidelines in your post.-->

## Text styling

- [X] The blog is written with [plain language](https://www.plainlanguage.gov/guidelines/) (where relevant).
- [X] If there are headers, they use the proper header tags in order to do so (with only one level-one header).
- [X] All links describe where they link to (for example, check the [Quansight labs website](https://labs.quansight.org/)).
- [X] Any kind of styling that the author uses (for example, bold for emphasis) is consistent throughout the blog.

## Non-text contents

- [X] Blog post featured image is in PNG or JPEG format, **not** SVG.
- [X] All content is represented as text (for example, images need alt text and videos need captions or descriptive transcripts).
- [X] If there are emojis, there are not more than three in a row.
- [ ] Don't use [flashing gifs or videos](https://www.w3.org/TR/UNDERSTANDING-WCAG20/seizure-does-not-violate.html). **One of the videos I've included does include a slow fade-in fade-out sequence, so would appreciate some guidence on if this is acceptable. If it's not then I'm very happy to update the video.
- [X] If it were to be read as plain text, the blog still makes sense and no information is missing.

## Current issues
The section on the array API and arbitrary regions of integration depends on not yet merged PRs.

At the moment, I'm including block-level math by wrapping it in `$$ ... $$` and inline math by wrapping it in `$ ... $`. However, this is causing issues when I try to build the site locally. If it is a simple formula, then the math will be displayed but not rendered (I can still see the `$$` on the page). If it's more complicated, then I get MDX errors like:

```
error [Error: [next-mdx-remote] error compiling MDX:
Could not parse expression with acorn: Expecting Unicode escape sequence \uXXXX

  13 | \begin{aligned}
  14 | &\int^1_0 \frac{1}{1 + x + x^3} \, \text dx \approx 0.6303 \\\\
> 15 | &\int^1_0 \frac{1}{1 + x + x^3} \, \text dx = \frac{2 ^ \frac{4}{3}\cdot3 ^ \frac{5}{2} \left(\frac{\sqrt{31}}{2\cdot 3 ^ \frac{3}{2}} - \frac 1 2\right)\ln\Big(\Big| 2- \cdots}{\sqrt[3]{\sqrt{31} - 3 ^ \frac{3}{2}}\Big(\Big(-\frac{3 ^ \frac{3}{2} \sqrt[3]{29 \cdot 3 ^ \frac{3}{2}\cdot}}{\sqrt[3]{2}} \,\,\,\cdots}
     |                                                          ^
  16 | \end{aligned}
  17 | $$

More information: https://mdxjs.com/docs/troubleshooting-mdx]
error - pages/blog/[slug].tsx (137:9) @ getStaticProps
error - TypeError: Cannot read properties of undefined (reading 'meta')
    at getStaticProps (webpack-internal:///./pages/blog/[slug].tsx:177:130)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async Object.renderToHTML (/Users/olly/code/sites/Quansight-website/node_modules/next/dist/server/render.js:368:20)
    at async doRender (/Users/olly/code/sites/Quansight-website/node_modules/next/dist/server/base-server.js:908:38)
    at async cacheEntry.responseCache.get.isManualRevalidate.isManualRevalidate (/Users/olly/code/sites/Quansight-website/node_modules/next/dist/server/base-server.js:1013:28)
    at async /Users/olly/code/sites/Quansight-website/node_modules/next/dist/server/response-cache.js:69:36 {
  page: '/blog/[slug]'
}
  135 |   const footer = await getFooter(preview);
  136 |   const featuredPosts = await getPostsByCategory(
> 137 |     post.meta.category,
      |         ^
  138 |     post.slug,
  139 |     2,
  140 |     preview,
```

There is [an example blog post which includes math using this syntax](https://github.com/Quansight/Quansight-website/blob/3ad7d0b9850f7d7d902b1f7ecfc957e7937e8db3/examples/another-post.md?plain=1#L164) so I'm a bit stuck about what to do. Is there another option I need to enable?

I would also appreciate it if someone could check the videos to make sure they don't violate the W3 guidelines. The fade-in/fade-out is very slow.

Thanks!

ev-br commented on apps/labs/posts/multidimensional-integration-in-scipy.md:
```diff
@@ -0,0 +1,264 @@
+    f=lambda x_arr: np.ones(x_arr.size[0]),  # f is the constant 1 function
+    a_func=[-1, lambda x: -np.sqrt(1-x**2)],
+    b_func=[ 1, lambda x:  np.sqrt(1-x**2)],
+)
```
this part is stale, predates `region=`

melissawm commented on .vscode/settings.json:
```diff
@@ -3,7 +3,7 @@
   "editor.codeActionsOnSave": {
-    "source.fixAll.eslint": true,
-    "source.fixAll.format": true
+    "source.fixAll.eslint": "explicit",
+    "source.fixAll.format": "explicit"
```
Can you explain why this is needed?

ollybritton replied:
Oops, didn't mean to include that! I was running into issues where saving would cause weird auto-formatting issues with the math formulas.

melissawm commented on apps/labs/posts/multidimensional-integration-in-scipy.mdx:
````diff
@@ -0,0 +1,313 @@
+    ...
+```
+
+After 20 years of maturing as an open source project, SciPy now contains over 15 different functions for numerical integration, all of which address different but overlapping use-cases and functionality. My task for this internship was to address a small gap in these offerings: improving the support for multidimensional numerical integration ("cubature"), as initially discussed [in this SciPy iss ...
````
Maybe add a link to the scipy.integrate docs?

melissawm commented on apps/labs/posts/multidimensional-integration-in-scipy.mdx:
```diff
@@ -0,0 +1,313 @@
+
+Previously, there was no function in SciPy that let you efficiently integrate functions that were both multidimensional and array-valued at the same time. For example, `quad_vec` integrates array-valued univariate functions, while `nquad` handles scalar-valued multivariate functions. But there was nothing that let you handle integrands that were both.
+
+## A small example: solving a problem with `cubature`
```
I LOVE this section!

melissawm commented on apps/labs/posts/multidimensional-integration-in-scipy.mdx:
```diff
@@ -0,0 +1,313 @@
+
+## Array API support
+
+One feature of the new `cubature` function that would be hard to support with `dblquad` is that it also supports the array API standard. For a good primer on what this means, see [The Array API Standard in SciPy](https://labs.quansight.org/blog/scipy-array-api). In brief, a lot of SciPy's code relies heavily on NumPy, which provides the `array` data structure used to efficiently implement most of ...
```
```suggestion
One feature of the new `cubature` function that would be hard to support with `dblquad` is that it also supports the [array API standard](https://data-apis.org/array-api/). For a good primer on what this means, see [The Array API Standard in SciPy](https://labs.quansight.org/blog/scipy-array-api). In brief, a lot of SciPy's code relies heavily on NumPy, which provides the `array` data structure used to efficiently implement most of the algorithms SciPy offers. But a considerable amount of recent effort has meant that lots of SciPy functions are now array-agnostic – you can give these functions arrays from other compatible libraries, like PyTorch or CuPy, and they will also work:
```

melissawm commented on apps/labs/posts/multidimensional-integration-in-scipy.mdx:
```diff
@@ -0,0 +1,313 @@
+
+For example, the simplest version of trapezium rule evaluates the function at the start and end of an interval, and estimates that the area of the function is roughly the same as the area of the trapezium formed by connecting up these two points.
+
+![An illustration of the premise behind the trapezoid rule. A curve is show, and layered on top is a trapezium that begins and ends at the start and end of the curve.](basic-trapezoid-rule.png)
```
```suggestion
![An illustration of the premise behind the trapezoid rule. A curve is shown, and layered on top is a trapezium that begins and ends at the start and end of the curve.](basic-trapezoid-rule.png)
```

melissawm replied:
This image does not show up in the preview

melissawm reviewed (commented):
Thanks Olly - left a couple of comments but nothing major. Cheers!

melissawm commented on apps/labs/posts/multidimensional-integration-in-scipy.mdx:
```diff
@@ -0,0 +1,313 @@
+---
+title: 'Multdimensional integration in SciPy'
```
```suggestion
title: 'Multidimensional integration in SciPy'
```

melissawm reviewed (commented):
Your title image is not loading in the preview - can you check?

ev-br commented on apps/labs/posts/multidimensional-integration-in-scipy.mdx:
```diff
@@ -0,0 +1,313 @@
+&\int^1_0 \frac{1}{1 + x + x^3} \, \text dx = \frac{2 ^ \frac{4}{3}\cdot3 ^ \frac{5}{2} \left(\frac{\sqrt{31}}{2\cdot 3 ^ \frac{3}{2}} - \frac 1 2\right)\ln\Big(\Big| 2- \cdots}{\sqrt[3]{\sqrt{31} - 3 ^ \frac{3}{2}}\Big(\Big(-\frac{3 ^ \frac{3}{2} \sqrt[3]{29 \cdot 3 ^ \frac{3}{2}\cdot}}{\sqrt[3]{2}} \,\,\,\cdots}
+\end{aligned}
+$$
+*/}
```
Based on an example from code search on the repository, 
https://github.com/search?q=repo%3AQuansight%2FQuansight-website%20%5Cfrac&type=code

it looks like TeX math needs a `math::` admonition, not sure. @melissawm would you be able to give us a hand please? Olly's post definitely needs TeX math but we're not sure how to make it render.

ollybritton replied:
I think this is actually TeX inside a docstring included in a code block.

ev-br replied:
Indeed! So mu suggestion is useless, sorry for the noise! The question of how to use TeX math is still relevant though.

melissawm replied:
Yeah, I think this is a blog setting, I'm really not sure 😦 Maybe someone from the website team can help - @pavithraes ?

melissawm reviewed (commented):
Make sure your videos also have alt-text 🙏

ollybritton commented:
I've just gone through and un-commented and escaped all the math. It's displaying the math as plain text and isn't causing any issues actually building the blog, at least on my computer.

If the script tag for MathJax/KaTeX were included on the page, it should then render the math correctly. I had a quick look at doing this, but as far as I can see doing this involves changing the blog internals rather than something like enabling a flag in the frontmatter. I tried directly including a script tag which would load MathJax from a CDN, but this also doesn't work (since it's probably not secure!). Hopefully someone with more experience will be able to enable math rendering?

I've also updated one of the formulas so that it is consistent with the rest of the notation in the blog post, namely that functions which return vectors are in bold.

And finally I've also changed the hero and featured images for this post to JPGs and modified them slightly so that they appear less pixelated.

melissawm commented:
Preview looks great, can we remove it from draft mode @ollybritton ?

ollybritton commented:
@melissawm Yes, I just made one last change to remove some bold text but I'm happy for it to be removed from draft mode now. Thanks for sorting out the math rendering!

melissawm commented:
@ev-br how does this look to you?

ev-br reviewed (approved):
LGTM

melissawm commented:
Let's do this, thanks all!
